vitis/hlsBs/man1/hlsComp.1
.TH hlsComp 1 "May 2026" "v1.0" "HLS Command Manual"
.SH NAME
hlsComp \- cleans, creates, removes, and lists components in the workspace from
configuration files

.SH SYNOPSIS
.B hlsComp [COMPONENTS] [OPTIONS]

.SH DESCRIPTION
.TP
.B hlsComp
This command manages the components in the workspace. It is mainly used
to add components for configuration files not created with \fBhlsCfg\fR, e.g.
configuration files created through the IDE or that already exist.

.SH COMPONENTS
A list of the names of the components, either as fully specified or
with wildcards.

.SH OPTIONS
.TP
\fB\-h, \-\-help\fR
Displays this message\

.TP
\fB\-\-verbose
Enables more extensive output which is useful for diagnostics.  The default
is 'False'.

.TP
\fB\-w, \-\-workspace = \fIWORKSPACE\fR
The target workspace. Note that for existing projects, the workspace is almost
all the information that is needed.  If there is no need to categorize
components as \fBexisting\fR, \fBmissing\fR, etc, no project file or a very
minimal one just specifying the workspace is needed.

.TP
\fB\-\-targets=...\fR
A list of component names.  Functionally this acts the same as the positional
\fBCOMPONENTS\fR arguments, but due to shell's lack of a formal binding syntax
for lists and the shell's handling of wildcards for positional arguments this
is more explicit and reliable.

.TP
\fB\-\-configurations=...\fR
A list of configuration files, either explicitly with full path names or with
wildcards. This defines the complete set of known configuration files and their
corresponding components. It is the basis of categorizing components as
\fBexisting\fR, \fBmissing\fR or \fBcruft\fR.

.TP
\fB\-\-components=...\fR
The name(s) of the components to create.  This defaults to naming the component
after the configuration file name. This provides a practical means of specifying
the component name when multiple configuration files are provided. The default
is essentially \fB{cfg_name}\fR, but can be overridden with this option. See the
examples.

.TP
\fB\-\-project = \fIPROJECT_FILE\fR
A project file is the standard method used by all the \fBhls\fR build utilities
of capturing the stable command line options in a file that can be saved in GIT.
For existing projects, the formality of a project file is not necessary. There
are simpler ways. Only for more complex projects is a project file needed. It is
mainly for consistency with other \fBhls\fR build utilities. Default is the
environment variable \fIHLSBS_PROJECT\fR.

.TP
\fB\-\-no-project
This prevents the use of any existing project file which may or may not exist.
Typically used when one wishes to specify all the relevant parameters (in this
utility these are \fBworkspace\fR and \fBconfigurations\fR and, possibly
\fBcomponents\fR if one wishes something other than the default. A common usage is
in one-off experimentation.

.SH ACTIONS
.TP
\fB\-\-clean\fR = (existing, missing, cruft, kruft, Kruft, all)
Cleans, i.e. removes a category of components. The default category is
\fBexisting\fR.  Given the heuristic nature of files categorized as
\fBcruft\fR it is recommended to first do either \fB--list\fR or \fB--dry-run\fR.

.TP
\fB\-\-create\fR = (existing, missing, cruft, kruft, Kruft, all)
Creates the component files. While any of the categories may be specified,
only \fBmissing\fR makes practical sense and is the default.

.TP
\fB\-\-list\fR = (existing, missing, cruft, kruft, Kruft, all)
Lists the status of the specified categories of component files. The default
is \fIall\fR, i.e. \fBexisting\fR, \fBmissing\fR and \fBcruft\fR.

.TP
\fB\-\-replace\fR = (existing, missing, cruft, kruft, Kruft, all)
Replaces the components. The default is \fBexisting\fR. Since this is effectively
a --\fBremove\fR and --\fBcreate\fR, use with caution.  As a bow to convenience,
specifying \fBmissing\fR will also create any missing component. See the
examples.

.TP
\fB\-\-dry-run\fR
Only reports the actions the command would do, but does not run the command.
Particularly useful and recommended for destructive commands, e.g. \fB--clean\fR
and \fB--replace\fR.


.SH Project File
A project file is used to capture values of various options that could be
specified on the command line. This saves both typing and remembering what they
are. Given that \fBhlsComp\fR is generally used in very simple existing projects,
no project file or a very minimal one is needed specifying
    1. The workspace
    2. The configuration file(s)
    3. The mapping of a configuration file name to a component


.SH Component Selection
Which components are selected for a given action is by the intersection of the
positional component list (and its equivalent \fB--targets\fR) and the component's
current state. The current state is one of
    1. Existing, the component exists and is consistent with a configuration file
    2. Missing,  The configuration file does not have a corresponding component
    3. Cruft,    A component found in the workspace or its reference to its
                 parent configuration file is inconsistent with the
                 configuration file specification.
Each action can specify a category or categories to target. For \fB--list\fR
all categories are possible. For others only certain categories make sense. For
example, only \fBexisting\fR makes sense for \fB--replace\fR.  For the most part
the utilities permit any combination and warn one against using inappropriate ones.


.SH Overriding with Command Line Options
The \fBhls\fR utilities are constructed so parameters can be specified at 3
levels depending on their permanency
    1. In the project file
    2. Directly overridden on the command line
    3. With parameter files on the command line

The \fBproject file\fR is the most permanent and \fBoverriding\fR directly on
the command line the least. The disadvantage of overriding directly on the
command line is that, except for command recall, they are lost to history and even
then, they may be seen as just a random collection of options.

The third option offers something in between, allowing one to capture
command line options in a file.  This allows the options some degree of
permanency and, since it is a file, it can be given a meaningful name. The
disadvantage is, due to the complexity in some utilities, not all options offered
in the project file are available.

The command line syntax for a parameter file is \fI@<MY_OPTIONS>\fR, i.e. the
file name preceded by an \fB@\fR. The file is trivially formatted, one option
(no spaces) per line. These act just as if the contents had been placed in
the command line where the parameter file is.
See the example below.

.SH Examples:

  Example 1
  $ hlsComp $MY_PROJECT/test.cfg --create \
           --workspace=$MY_PROJECT/ws/{vitis_version}
    Creates the component \fBtest\fR in the workspace. Note no project
    file is needed in this case

  Example 2
  $ hlsComp $MY_PROJECT/*.cfg --create
            --workspace=$MY_PROJECT/ws/{vitis_version}
    Creates any missing components in the workspace named after the
    configuration files satisfying the wildcard specification.

  Example 3
  $ hlsComp --create
            --workspace=$MY_PROJECT/ws/{vitis_version}
            --configurations=/MY_PROJECT/cfg/*.cfg
            --components=my-{cfg_name}

    Similar to the previous command, but adds the \fImy-\fR prefix to the
    component name.  Adding a prefix may be useful in one-off experimentation.

  Example 4
  $ hlsComp --workspace=$MY_PROJECT/ws/{vitis_version}
            --list=missing
            --configurations=$MY_PROJECT/cfg/*.cfg

    Will list only the missing components.

  Example 5
  $ hlsComp '*' --create @MyOption.txt --workspace=$EXP/ws/{vitis_version}

  Assuming \fBMyOptions.txt\fR contains
       --workspace=$MY_PROJECT/ws/{vitis_version}
       --configurations=$MY_PROJECT/cfg/*.cfg
       --components=my-{cfg_name}

  This duplicates Example 3.  The ordering is important, but in the sense
  that the last instance of a parameter supersedes previous ones.

  Here the explict --workspace=$EXP/ws/{vitis_version} overrides both
  the definition in the project file and the MyOption.txt parameter file
  by virtue of being last.

  This is handy for one-off experimentation.

.SH AUTHOR
Written by JJRussell


vitis/hlsBs/vitispy/hlsComp.py
# ----------------------------------------------------------------------------
# This file is part of the 'SLAC Firmware Standard Library'. It is subject to
# the license terms in the LICENSE.txt file found in the top-level directory
# of this distribution and at:
#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
# No part of the 'SLAC Firmware Standard Library', including this file, may be
# copied, modified, propagated, or distributed except according to the terms
# contained in the LICENSE.txt file.
# ----------------------------------------------------------------------------

# ------------------------------------------------------------------------------
'''
    Creates/Removes/Queries HLS Components

    Args:
       @project_file  :  Gives the default workspace
       -w, --workspace:  Specifies the workspace, overriding the default

'''
# ------------------------------------------------------------------------------


import os
import sys
import argparse
import shutil
import runpy


# -------------------------------------------------------
# Add the VITIS HLS python paths, remove nonexistent ones
# ------------------------------------------
runpy.run_path(os.getenv('HLSBS_IMPORT_PYPATHS'),
               run_name='add_paths ' + str(sys.path))

from vitispy.componentInfo import ComponentInfo
from vitispy.category import Category
from vitispy.targets import Targets
from vitispy.project import Project
from vitispy.dry_run import DryRun
from vitispy.action import Action
from vitispy.workspace import Workspace
from vitispy.printer import Printer
import vitispy.files as files
from vitispy.manpage import display_manpage
import vitis

# ==============================================================================
#
# This creates, replaces, cleans, lists components in a workspace
#
# ==============================================================================


# ==============================================================================
# BEGIN: Local methods
# ------------------------------------------------------------------------------
def add_arguments(parser):
    Project.  add_arguments(parser)
    Workspace.add_arguments(parser)
    DryRun.   add_arguments(parser)
    Action.   add_arguments(parser, 'Components')

    parser.add_argument('-h',
                        '--help',
                        action='store_true',
                        help='Show custom help')

    parser.add_argument('--cfg-root',
                        default=None,
                        dest='cfg_root',
                        help='Configuration files root directory')

    parser.add_argument('parameters',
                        help='List of targets/components',
                        nargs='*',
                        action='store',
                        default=None)

    parser.add_argument('--verbose',
                        action='store_true',
                        default=False)

    parser.add_argument('--targets',
                        help='Inputs are targets',
                        nargs='?',
                        const='*_no_value_*',
                        default=None)

    parser.add_argument('--components',
                        help='Manual entry of components to create',
                        nargs='?',
                        const='*_no_value_*',
                        default=None)

    parser.add_argument('--configurations',
                        help='Manual entry of target configurations',
                        nargs='?',
                        const='*_no_value_*',
                        default=None)

    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def get_opts(action, args):
    if action & Action.List:
        label = 'Listing'
        opts = Category.categorize(args.list, Category.All)

    elif action & Action.Clean:
        label = 'Cleaning'
        opts = Category.categorize(args.clean, Category.Existing)

    elif action & Action.Replace:
        label = 'Replacing'
        opts = Category.categorize(args.replace, Category.Existing)

    elif action & Action.Create:
        label = 'Creating'
        opts = Category.categorize(args.create, Category.Missing)

    else:
        label = 'Creating'
        args.create = ['missing']
        opts = Category.categorize(args.create, Category.Missing)

    return label, opts
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def get_client(action, workspace, targets):
    if action & (Action.Create | Action.Replace):
        for target in targets:
            cfg_exists = os.path.isfile(target.cfg_path)

            # Create create component with no cfg
            if not cfg_exists:
                continue

            cmp_exists = os.path.isdir(target.cmp_path)

            # Can't create if cmp already exists
            if action & Action.Create and cmp_exists:
                continue

            # Need to create the client
            client = vitis.create_client()
            Workspace.set(client, workspace)
            return client
    return None
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noWorkspaceError():
    print(
        '''
    ERROR: No workspace was provided
           add --workspace=<ws>
    ''',
        file=sys.stderr)
    return -1
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def nonExistentWorkspace(workspace):
    print(
        "ERROR: Workspace does not exist\n"
        "       To avoid typos causing spurious creation, use hlsWs to create it\n"
        "\n"
        "  -->  " + workspace + "\n", file=sys.stderr)
    return -1
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noCompleteCleanWarning():
    # Error message when an implicit '*' is used when cleaning
    print(
        '''
    WARNING: To avoid cleaning more than intended, hlsComp requires that
             positional args or --targets='*' must be specified, e.g.
               $ hlsCfg '*' --clean             or
               $ hlsCfg --targets='*'
    ''',
        file=sys.stderr)
    return -1
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def cmpExistsError(printer):
    printer.itemPlain('', "ERROR: component already exists, use --replace", '*')
    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noCfgError(printer):
    printer.itemPlain('', 'ERROR: cannot replace component',          sep='*')
    printer.itemPlain('', '       configuration file does not exist', sep=' ')
    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noConfigurationList():
    print('''
           ERROR: No configuration target list provided
           ''',
          file=sys.stderr)
    return -1
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def noCompError(printer, workspace, tgtList):
    all = sorted(files.get_components(workspace, '*'))

    if not all:
        printer.line("ERROR: No components found in workspace")
        printer.line(f"       {workspace}")

    else:
        printer.line(f"ERROR: No components found matching <{tgtList}>")
        printer.line("Candidates are:")

        idx = 1
        for component in all:
            cmp_name = os.path.splitext(os.path.split(component)[1])[0]
            printer.item(idx, cmp_name, None)
            idx += 1
    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
class Executer:
    def __init__(self,
                 action,
                 printer,
                 verbose,
                 opts,
                 workspace,
                 categories,
                 dry_run):

        self.action = action
        self.printer = printer
        self.verbose = verbose
        self.opts = opts
        self.workspace = workspace
        self.categories = categories
        self.dry_run = dry_run
        return
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    def byCategories(self, client):

        printer = self.printer

        # ------------------------------------------------------
        # Only print dry run if it causes some 'write' activity
        # ------------------------------------------------------
        if self.dry_run is not None and self.dry_run:
            self.printer.line('Dry Run', '<-- NOTE')
        self.printer.separator('-')

        lf = False
        if self.opts & Category.Existing:
            printer.line('Existing')
            lf = self.execute(client, self.categories.existing, True)

        if self.opts & Category.Missing:
            if lf:
                print()
            self.printer.line('Missing Component')
            lf = self.execute(client, self.categories.missing, True)

        if self.opts & Category.Cruft:
            if lf:
                print()
            lf = False
            nlo = len(self.categories.cruft_lo)
            nhi = len(self.categories.cruft_hi)

            self.printer.line('No Configuration File')
            if (nlo == 0) and (nhi == 0):
                self.printer.itemPlain("None", "")

            if self.opts & Category.CruftHi and nhi:
                if lf:
                    print()
                lf = self.execute(client, self.categories.cruft_hi, False)

            if self.opts & Category.CruftLo and nlo:
                if lf:
                    print()
                lf = self.execute(client, self.categories.cruft_lo, False)

        return
    # --------------------------------------------------------------------------

    # --------------------------------------------------------------------------

    def execute(self,
                client,
                targets,
                print_none):

        printer = self.printer

        if len(targets) == 0 and print_none:
            printer.itemPlain('None', None, None)
            return True

        idx = 0
        lf = True
        for target in targets:
            cfg_path = target.cfg_path
            cmp_path = target.cmp_path
            cmp_name = target.cmp_name

            cfg_exists = os.path.isfile(cfg_path) if cfg_path else False
            cmp_exists = os.path.isdir(cmp_path) if cmp_path else False
            cfg_sep = ':' if cfg_exists else '*'
            cmp_sep = ':' if cmp_exists else '*'

            if idx != 0 and lf:
                print()
            idx += 1
            name = cmp_name
            if cmp_name is None:
                name = ("Component not present in workspace, "
                        "check cfg file as a possible stray")

            printer.item(idx, 'Component', name, cmp_sep)
            if (cfg_path and self.verbose) or cmp_name is None:
                printer.itemPlain('Configuration', cfg_path, cfg_sep)
            else:
                lf = False

            # ACTION = LIST
            if self.action & Action.List:
                # The list has effectively already been performed
                continue

            # ACTION = CLEAN
            if self.action & Action.Clean:
                if cmp_exists:
                    if not self.dry_run:
                        shutil.rmtree(target.cmp_path)

            # ACTION = CREATE
            if self.action & Action.Create:
                if cmp_exists:
                    cmpExistsError(printer)
                elif not cfg_exists:
                    noCfgError(printer)
                else:
                    if self.dry_run:
                        continue
                    client.create_hls_component(name=cmp_name,
                                                cfg_file=[cfg_path],
                                                template="empty_hls_component")

            # ACTION = REPLACE
            if self.action & Action.Replace:
                if not cfg_exists:
                    noCfgError(printer)
                else:
                    if self.dry_run:
                        continue

                    client.delete_component(name=cmp_name)
                    client.create_hls_component(name=cmp_name,
                                                cfg_file=[cfg_path],
                                                template="empty_hls_component")
        return True
    # --------------------------------------------------------------------------

# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def merge(xl, yl):
    '''
    Merges xl, yl into a list, where xl,yl can either be a list or a comma
    separated string.
    '''
    # --------------------------------------------------------------------------
    lst = []
    if xl:
        # Check if XL is a list
        if isinstance(xl, list):
            for x in xl:
                lst += x.split(',')
        else:
            lst = xl.split(',')

    if yl:
        # Check if YL is a list
        if isinstance(yl, list):
            for y in yl:
                lst += y.split(',')
        else:
            lst += yl.split(',')

    return lst
# ------------------------------------------------------------------------------
# END  : Local methods
# ==============================================================================


# ==============================================================================
# BEGIN: Main execution
# ------------------------------------------------------------------------------
def doit():

    # -------------------------------
    # Get the command line parameters
    # -------------------------------
    parser = argparse.ArgumentParser(add_help=False,
                                     fromfile_prefix_chars='@')
    add_arguments(parser)

    args = parser.parse_args()
    if args.help:
        display_manpage(__file__)
        exit(0)

    action = Action(args).action
    dry_run = args.dry_run
    cmpList = merge(args.parameters, args.targets)

    # -------------------------------------------
    # Check if target list of components is empty
    # -------------------------------------------
    if not cmpList:
        if action & Action.Clean:
            return noCompleteCleanWarning()
        else:
            cmpList = ['*']

    label, opts = get_opts(action, args)
    printer = Printer(20, 78, 15)
    noCfgAction = False
    full = not args.configurations and args.project is not None
    needs = (Project.Need.Root |
             Project.Need.Workspace |
             Project.Need.Products_Root |
             Project.Need.Build_Root |
             Project.Need.Cfg_Root)
    project_files = args.project if full else None

    project = Project(needs,
                      project_files,
                      args.root,
                      args.products_root,
                      args.build_root,
                      args.workspace,
                      args.cfg_root,
                      None)

    if project.error:
        status = project.report(needs)
        return status

    if not full:
        # -----------------
        # Vet the arguments
        # -----------------
        if not args.configurations or len(args.configurations) == 0:
            if not action & (Action.List | Action.Clean):
                return noConfigurationList()
            else:
                noCfgAction = True
    else:
        full = hasattr(project, 'get_products')

        # -------------------------------------------------------------
        # Override any project specific defaults with command line ones
        # -------------------------------------------------------------
        project.replace(args)

    # -----------------------------------------------------------------
    # Demand the workspace exists
    # Automatically creating is dangerous,
    # Doing so can unknowingly create and populate a mistyped workspace
    # which can easily go unnoticed
    # -----------------------------------------------------------------
    workspace = project.workspace
    exists = os.path.isdir(workspace)
    if not exists:
        return nonExistentWorkspace(workspace)

    # ---------------------------------------------------
    # Check if doing list or clean with no configurations
    # ---------------------------------------------------
    if noCfgAction:
        opts &= ~Category.Missing
        categories = Category.categorizeByComponent(workspace, cmpList)

    # Check if doing the cheap version, i.e. do project products
    elif not full:
        if not args.components:
            args.components = '{cfg_name}'

        configurations = os.path.expandvars(args.configurations)
        components = os.path.expandvars(args.components)
        categories = Category.getCategories(workspace,
                                            configurations,
                                            components,
                                            cmpList)
    else:

        # ---------------------------------------
        # To do create/replace, need the products
        # --------------------------------------

        project.get_products()
        targets = Targets(workspace, project.products).targets
        targets = Targets.filter(targets, cmpList).accepts

        if not targets:
            msg = label + " Components"
            return Targets.no_candidate_targets(targets, msg, project)

        categories = Category.byComponents(targets)

    # ---------------------------------------------
    # Create HLS client only if necessary.
    # It is time wise expensive and dumps confusing
    # and unnecessary crap into the output
    # ---------------------------------------------
    client = None
    if not dry_run:
        if action & Action.Create:
            client = get_client(action,
                                workspace,
                                categories.missing)
        elif action & Action.Replace:
            client = get_client(action,
                                workspace,
                                categories.existing)

    if noCfgAction:
        printer.header(label + ' Components')
        printer.line('Workspace',      workspace)
        printer.line('Component Filter', cmpList)

    elif not full:
        printer.header(label + ' Components')
        printer.line('Workspace', workspace)
        printer.line('Configurations', args.configurations)
        printer.line('Component Template', args.components)
        printer.line('Component Filter', cmpList)
    else:
        printer.header(label + ' Components')
        printer.line('Workspace',      workspace)
        printer.line('Component Filter', cmpList)

    if not categories:
        printer.footer()
        return -1
    else:
        executer = Executer(action,
                            printer,
                            args.verbose,
                            opts,
                            workspace,
                            categories,
                            dry_run)
        executer.byCategories(client)
        printer.footer()

        return 0
# ----------------------------------------------------------------------


# ----------------------------------------------------------------------
if (__name__ == '__main__'):
    status = doit()
    sys.exit(status)
# ----------------------------------------------------------------------
